### PR TITLE
graph and function utilities

### PR DESCRIPTION
**Pull request by xlauko**
implements:

- basic graph concepts
- depth first search generic algorithm
- topological sort
- setup coroutines
- add hashing library
- add memoization wrapper

### CODE REVIEW

File: core/include/gap/core/graph.hpp [hunk 1]
```diff
@@ -0,0 +1,115 @@
+// Copyright 2022-present, Trail of Bits, Inc.
+
+#pragma once
+
+#ifdef GAP_ENABLE_COROUTINES
+
+    #include <gap/core/generator.hpp>
```
Comment (lkorenc) [inline core/include/gap/core/graph.hpp:7]: indent?

File: core/include/gap/core/memoize.hpp [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+#include <tuple>
+#include <unordered_map>
+#include <utility>
+
+namespace gap
+{
+    namespace detail
+    {
+        template< typename T, typename enable_t = void >
+        struct function_signature;
```
Comment (lkorenc) [inline core/include/gap/core/memoize.hpp:17]: Pull this out into some `type_utils` header? Sounds it may be useful in other places as well.

File: core/include/gap/core/memoize.hpp [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+        using return_reference = const result_type &;
+
+        using hash_type  = gap::hash< tuple_type >;
+        using cache_type = std::unordered_map< tuple_type, result_type, hash_type >;
+
+        explicit memoizer(function_type &&fn) : fn(std::forward< function_type >(fn)) {}
+
+        template< typename... call_args_t >
+        return_reference operator()(call_args_t &&...args) {
+            return call< std::conditional_t<
```
Comment (lkorenc) [inline core/include/gap/core/memoize.hpp:103]: hard to parse, can it be extracted and named?

File: core/include/gap/core/memoize.hpp [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+            return cache.count(tuple);
+        }
+
+      private:
+        template< typename... call_args_t >
+        return_reference call(call_args_t &&...args) {
+            auto tuple = std::forward_as_tuple(std::forward< call_args_t >(args)...);
+            if (auto res = cache.find(tuple); res != cache.end()) {
+                return res->second;
+            } else {
```
Comment (lkorenc) [inline core/include/gap/core/memoize.hpp:122]: `else-after-return` I think here every simplification of the code helps.

**Review by lkorenc — APPROVED** [review]

File: cmake/gapConfig.cmake.in [hunk 5]
```diff
@@ -6,6 +6,9 @@ include(CMakeFindDependencyMacro)
 list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_LIST_DIR}")
 
 if (@GAP_ENABLE_COROUTINES@)
+    set( CMAKE_CXX_STANDARD 20 )
+    set( CMAKE_CXX_EXTENSIONS OFF )
```
[the remark this answers is not in the document]
Comment (xlauko) [reply cmake/gapConfig.cmake.in:10]: for some magic reason `cxx_std_20` did not work :(

[the remark this answers is not in the document]
Comment (pgoodman) [reply cmake/gapConfig.cmake.in:10]: I think you should be setting `cxx_std_20` as an `INTERFACE_COMPILE_FEATURES` to `gap-settings` using `set_target_properties`.

File: core/include/gap/core/memoize.hpp (hunk 2 above)
Comment (xlauko) [reply core/include/gap/core/memoize.hpp]: resolved in 46c925ccf574e5203dfd126106a679145238f102
